include/picongpu/plugins/misc/rangeToWindow.hpp

```diff
@@ -0,0 +1,115 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation/control/Window.hpp"
+
+#include <pmacc/mappings/simulation/SubGrid.hpp>
+
+#include <algorithm>
+#include <string>
+
+
+namespace picongpu
+{
+    namespace plugins
+    {
+        namespace misc
+        {
+            /** Calculate window offsets and sizes based on a user given range description.
+             *
+             * @tparam T_dim number of dimensions the simulation is compiled for
+             * @param dataDomain PIConGPU data domain description
+             * @param inputWindow Selected window within the data domain e.g. the selected moving window.
+             * @param selectedRange String with range description for each dimension. Dimensions are separated by
+             *                      comma. The range for a dimension is of the form BEGIN:END where BEGIN is included
+             *                      and END excluded from the domain.
+             *                      If the begin or end of the range is outside of the data domain the values will be
+             *                      cropped to fit the data domain.
+             * @return Window description with local and global information.
+             */
+            template<uint32_t T_dim>
+            inline Window rangeToWindow(
+                pmacc::SubGrid<T_dim> const& dataDomain,
+                Window const& inputWindow,
+                std::string const& selectedRange)
+            {
+                const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+
+                auto parsedSlice = pmacc::pluginSystem::toTimeSlice(selectedRange);
+
+                Window resultWindow = inputWindow;
+                // Parse the ranges only for dimensions the user is providing.
+                for(uint32_t d = 0; d < parsedSlice.size() && d < simDim; ++d)
+                {
+                    auto range = parsedSlice[d];
+                    uint32_t const windowOffset = inputWindow.globalDimensions.offset[d];
+                    uint32_t const windowSize = inputWindow.globalDimensions.size[d];
+                    uint32_t const windowEnd = windowOffset + windowSize;
+                    uint32_t const rangeBegin = range.values[0];
+                    // Crop the user given end range to the end of the input window.
+                    uint32_t const rangeEnd = std::min(range.values[1], windowEnd);
```

Comment:
Is this correct? Isn't here `windowEnd` as total coordinates and `range.values[1]` in global ones?

Reply:
Discussed offline, all fine here, I didn't understand the code.